index.mjs

```diff
@@ -0,0 +1,165 @@
+ * @param {string|undefined} source
+ * @returns {Diagnostic}
+ */
+function vfileMessageToDiagnostic(message, source) {
+  const diagnostic = Diagnostic.create(
+    unistLocationToLSPRange(message.position),
+    message.reason,
+    message.fatal ? DiagnosticSeverity.Error : DiagnosticSeverity.Warning,
+    message.ruleId
+      ? message.source
+        ? message.source + ':' + message.ruleId
+        : message.ruleId
+      : message.source || undefined,
+    source
+  )
+  if (message.url) {
+    diagnostic.codeDescription = {href: message.url}
+  }
+
+  return diagnostic
+}
+
+/**
+ * Convert language server protocol text document to a vfile.
+ *
+ * @param {TextDocument} document
+ * @returns {import('vfile').VFile}
+ */
+function lspDocumentToVfile(document) {
+  return new VFile({
+    // VFile expects a file path or file URL object, but LSP provides a file URI
+    // as a string.
+    path: new URL(document.uri),
+    value: document.getText()
+  })
+}
+
+/**
+ * @param {import('vscode-languageserver').Connection} connection
+ * @param {import('vscode-languageserver').TextDocuments<TextDocument>} documents
+ * @param {import('unified-engine').Options['plugins']} plugins
+ * @param {string} prefix
+ */
+function initUnifiedLanguageServer(connection, documents, prefix, plugins) {
+  /**
+   * Process various LSP text documents using unified and send back the
+   * resulting messages as diagnostics.
+   *
+   * @param {TextDocument[]} textDocuments
+   */
+  function processDocuments(...textDocuments) {
+    engine(
+      {
+        files: textDocuments.map((document) => lspDocumentToVfile(document)),
+        ignoreName: '.' + prefix + 'ignore',
+        packageField: prefix + 'Config',
+        pluginPrefix: prefix,
+        plugins,
+        processor: unified(),
+        rcName: '.' + prefix + 'rc',
+        silentlyIgnore: true
+      },
+      (error, code, context) => {
+        if (error) {
+          console.error(error)
+        }
+
+        if (!context?.files) {
+          return
+        }
+
+        for (const file of context.files) {
+          connection.sendDiagnostics({
+            // VFile uses a file path, but LSP expects a file URL as a string.
+            uri: String(pathToFileURL(file.path)),
+            diagnostics: file.messages.map((message) =>
+              vfileMessageToDiagnostic(message, prefix)
+            )
+          })
+        }
+      }
+    )
+  }
+
+  connection.onInitialize(() => ({
+    capabilities: {
+      textDocumentSync: TextDocumentSyncKind.Full
+    }
+  }))
+
+  documents.onDidChangeContent(({document}) => {
+    processDocuments(document)
+  })
+
+  connection.onDidChangeConfiguration(() => {
+    processDocuments(...documents.all())
+  })
+}
+
+const connection = createConnection(ProposedFeatures.all)
+const documents = new TextDocuments(TextDocument)
+
+initUnifiedLanguageServer(connection, documents, 'remark', [
+  'remark-parse',
+  'remark-stringify'
+])
```

Comment:
The manual creation of `connection` and `document` is necessary for testing purposes, but it may be removed for a public interface.

Now that I have a better idea of how this should work, I can think of 2 options for a public interface. Personally I prefer option 2.

### Option 1: Provide a CLI for the unified language server

This means the end user will have to specify some options.

```sh
unified-language-server --prefix remark --plugin remark-parse --plugin remark-stringify --stdio
```

### Option 2: Provide a NodeJS API

This means for every unified ecosystem a language server needs to be created. This is similar to `unified-args`.

Creating `remark-language-server` will be as simple as:

```js
import {initUnifiedLanguageServer} from 'unified-language-server'

initUnifiedLanguageServer('remark', [
  'remark-parse',
  'remark-stringify'
])
```

End users will simply run:

```sh
remark-language-server --stdio
```

Reply:
There two options aren't strictly mutually exclusive.
If we start with option 2, a NodeJS API, we could build option 1 a CLI on top of that later if wanted.

Reply:
How are language servers used? Who is using this already, and how (or what projects do similar stuff that should switch to using this project?)

Reply:
> How are language servers used?

I mostly used language servers through VSCode.
Which wraps usually wraps the language server as a VSCode extension.
I'm not as familiar with how it is used with other editors.

> what projects do similar stuff that should switch to using this project?

I think there are some MDX language servers which could be migrated to use this.
And the remark, and incoming rehype (https://github.com/retextjs/retext/issues/57), language servers would likely need to be updated to leverage the new structure/features

Reply:
[Many editors](https://microsoft.github.io/language-server-protocol/implementors/tools/) support this using a minimal amount of code (if any). Based on this language server, it should be fairly straight forward to integrate unified into any of those editors.

Language servers provide many features, but the ones that make sense for unified are linting and formatting.

My goal is to make [vscode-remark](https://github.com/remarkjs/vscode-remark) use this language server. A similar plugin could be created for rehype, redot, and possibly others.

Based on the fact that [unified-engine-atom](https://github.com/unifiedjs/unified-engine-atom) still uses CJS, it’s outdated. That project can probably be discontinued in favor of plugins based on this language server.

Reply:
I like fewer things to exist so I'd vote for there being one executable that can be edited to fit multiple roles. One language-server that can both to plain-text, HTML, markdown, asciidoc, .. with the help of plugins.

Reply:
### Option 3: Support language server settings

Language servers support configuration options. This can be leveraged to configure `unified-engine`.

```viml
if (executable('unified-language-server'))
  au User lsp_setup call lsp#register_server({
  \ 'name': 'remark',
  \ 'cmd': {server_info->['unified-language-server', '--stdio']},
  \ 'allowlist': ['markdown'],
  \ 'config': {
  \   'defaultSource': 'remark',
  \   'ignoreName': '.remarkignore',
  \   'packageField': '.remarkConfig',
  \   'pluginPrefix': 'remark',
  \   'plugins': ['remark-parse', 'remark-stringify'],
  \   'rcName': '.remarkrc'
  \ })
endif
```

---

I still think option 2 is the way to go. This requires no additional configuration for end users, making it the simplest to use and leaving the least room for users to misconfigure it.

However, option 3 would be nice for some use cases. For example debugging `unified-language-server` or fiddling with a new ecosysem based on unified.

These options don’t have to exclude each other.

Reply:
I couldn’t get loading configuration to work, so I suggest to create an issue for it so it can be implemented later.